### ISSUE TITLE
UniValue should only parse the first JSON construct of a line

### ISSUE THREAD

**Issue by domob1812** [bitcoin/bitcoin#6558]
The current parser `UniValue::read` continues until it gets to the end of the parsed string (`JTOK_NONE`) or until an error occurs while lexing (`JTOK_ERR`).  If the string contains, for instance, multiple top-level objects or arrays, they are all parsed and only the last is saved in the `UniValue` object.  On the other hand, parsing fails if a valid object is followed by garbage.  (While parsing does *not* fail, surprisingly, if the string starts with a `JTOK_ERR`.)

I think this behaviour is highly confusing and should be changed.  I'm willing to work on this myself, but would like to discuss it before starting.  My suggestion is the following:

Reading from a string should only parse until a full JSON construct is found (object or array as now, or a single number/string/keyword).  Afterwards, parsing should stop and discard any trailing parts of the string completely.  I want to add also the option to return (in an optional output argument) the position where parsing stopped, so that the caller may get everything in there or check that no unparsed content remains; I do not think that Bitcoin currently needs this functionality, but it is almost "for free" and may be useful in the future.

Does this sound like a good idea, or is there a problem with my plan that I missed?

**Comment by jonasschnelli** [comment]
I agree. IMO `UniValue::read` should only read the first top level object/array and if another object opens (`[` or `{`) (or garbage), it should abort the parsing and response with an error. This would be conform with the JSON standard.

Would be nice if you could try to PR a fix for this (please also provide unit tests).

**Comment by domob1812** [comment]
Yes, of course I will write the patch (as promised), including proper testing.

I personally think that a recursive implementation of `UniValue::read` would be easier to read than the current one with explicit stack.  Is it ok if I try to rewrite it as I think it is best, or should I try to keep the changes to a minimum?  I. e., basically just change the stopping criterion of the loop while keeping the explicit stack?

**Comment by dcousens** [comment]
Why do we have our own JSON parser anyway?

**Comment by jonasschnelli** [comment]
`UniValue` is a very flexible and lightweight JSON parser/encoder. We just moved from JSON Spirit to UniValue because it's more lightweight and reduces compile time.

**Comment by jonasschnelli** [comment]
@domob1812 I think keeping the changes at a minimum would be good. If you have a bigger changeset in mind, you might open a PR for that at the UniValue upstream (https://github.com/jgarzik/univalue).

**Comment by domob1812** [comment]
Alright, was just a thought.  I'll submit a minimal PR for Bitcoin and will think about upstream patches later.

**Comment by jgarzik** [comment]
Related:  PRs that sync upstream with latest Bitcoin Core appreciated...

**Comment by domob1812** [comment]
@jgarzik: Is there some "defined procedure" for merging upstream patches for UniValue down to Bitcoin Core?  E. g., do you do that regularly or should all patches be submitted against both repositories?  If there is some procedure, I can very well submit my work to your repository first and have it "flow down" from there.

**Comment by jgarzik** [comment]
@domob1812 No defined procedure.  Currently Bitcoin Core is ahead of "upstream", which needs to be remedied.

**Comment by jonasschnelli** [comment]
@domob1812: one way could be to improve/fix in this repository with focus on a small diff. Once merged, open a PR upstream and sync bitcoins univalue with upstream. Then you could make more conceptual changes upstream, once it's stable, open a PR in this repository with the upstream changes.

**Comment by domob1812** [comment]
Thanks for the input, this is actually what I have been thinking as well.  I'll do that.

**Comment by sipa** [comment]
@Jeff after fixing and synchronizing upstream, we can maybe start treating
univalue as a subtree like leveldb and secp256k1? There are utility scripts
now to verify that the included code matches the specifies commit in the
upstream repo etc.

**Comment by laanwj** [comment]
> I personally think that a recursive implementation of UniValue::read would be easier to read than the current one with explicit stack

An explicit stack is preferred here. Easier to keep bounds on memory usage (we could, for example, limit the max depth easily). Also: execution stacks can be shallow on some architectures especially those of threads.

**Comment by jgarzik** [comment]
+1 @sipa RE subtree
+1 @laanwj RE explicit stack

**Comment by jonasschnelli** [comment]
subtree make sense. I just synced upstream (https://github.com/jgarzik/univalue/commits/master). As soon as this is merged, i try to PR including UniValue over a subtree.
+1 for explicit stack... but maybe another PR, maybe upstream.
